### ISSUE TITLE
Can we set startTime after manifestparsed event is emitted?

### ISSUE THREAD

**Issue by caridley** [google/shaka-player#3434]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
yes - i found one related issue but it did not address exactly the same question
https://github.com/google/shaka-player/issues/1298 Seek After load does not work

**What version of Shaka Player are you using?**
3.1.0

**Please ask your question**
For VOD DAI content the bookmark positions that we use to resume playback of a VOD asset do not include the advertising periods because the number and duration of ads may be different for different playback sessions. Because of this we do not know the exact position at which to start playback until the the manifest has been parsed and we received ad meta-data via “timelineregionadded” events.

Currently we are calling load() with a startTime of zero and then after load completes seeking to the desired start position. When we do this the player loads init and media segments for position zero before we are able to start the seek, causing a small delay in stream startup.

Is there anyway for us to set the playback start time when the manifestloaded event is emitted?

**Comment by joeyparrish** [comment]
I don't believe we have an explicit feature for this, but I think it sounds reasonable.

What would be ideal, in my opinion, would be to avoid a new method and just allow the application to set `video.currentTime` from the `timelineregionadded` events before the [`streaming` event](https://shaka-player-demo.appspot.com/docs/api/shaka.Player.html#.event:StreamingEvent) fires.  That would be pretty natural, I think.  But it's possible that it might not work today, and it's also possible that there might be reasons it would be difficult to support.

What happens today when you set `video.currentTime` before the `streaming` event fires?  Does it work?  Is it ignored?  Does it confuse the player somehow?

**Comment by caridley** [comment]
I don't think it will work to set video.currentTime from the `timelineregionadded` event handler because `timelineregionadded` is emitted for each ad, and we can't compute the desired seek position until we have data for all of the ads. 

I have tried setting `video.currentTime` in the `manifestparsed` and `streaming` event handlers. It works, in that playback appears to starts at the requested position, but it does not prevent the streaming engine from loading unneeded init and media segments from the first period.

**Comment by caridley** [comment]
If I add this code to Player.onLoad_() 

```
    // Since we are about to start playback, we will lock in the start time as
    // something we are now depending on.
    has.startTime = wants.startTime;
    if(has.mediaElement.currentTime && has.mediaElement.currentTime > 0) {
      has.startTime = has.mediaElement.currentTime;
    }

```

and then set video.currentTime in my 'manifestparsed' handler I get the desired behavior. Playback starts at the requested position and the player does not load unneeded segment from the first period.

**Comment by joeyparrish** [comment]
That seems pretty reasonable at a glance.

You don't need a check for both currentTime && currentTime > 0.  It can't be negative, so checking > 0 should be enough.

**Comment by caridley** [comment]
@joeyparrish  I would like to submit a PR with the proposed change. Does this issue need to be converted into a bug or enhancement request?

**Comment by caridley** [comment]
@joeyparrish  I have found that I am unable to set the value of mediaElement.currentTime in the `manifestparsed` or `streaming` event handlers with Safari (it works with Chrome, Firefox and Edge). The value just doesn't stick. Can you think of any other way set a start time after the manifest has been parsed?

**Comment by joeyparrish** [comment]
That makes sense, because on Safari, Shaka Player is probably using Apple's native HLS for playback.  In this scenario, we don't get to parse the manifest or do any streaming ourselves.  With native HLS, we only control DRM licensing.

In this scenario, you may be able to use ready-state events from the video element.  For example:

```js
// This only fires for MediaSource-based playbacks:
player.addEventListener('manifestparsed', myInitialSeek);
// This fires for both MediaSource-based playbacks and native HLS playbacks:
video.addEventListener('loadedmetadata', myInitialSeek, {once: true});

await player.load('foo.m3u8');
```

Because `loadedmetadata` events fire for both kinds of playback, you may want your handler to return early if it has already fired on the other event type.

**Comment by caridley** [comment]
I ran into the Safari problem when creating an integration test using a DASH stream. The `manifestparsed` event is being fired but I am unable to set the `currentTime`. I can post a PR if that would be helpful.

I don't think we intend to use DASH on Safari, but I am concerned that this might be a problem for other webkit based environment.

**Comment by joeyparrish** [comment]
It's possible that WebKit requires you to wait for `loadedmetadata`, even for MediaSource.  Can you try using that event only, instead of `manifestparsed`?  That will fire when the first init segment is appended.

**Comment by caridley** [comment]
I'm not having much luck with updating mediaElement.currentTime in the `loadedmetadata` event (which is emitted by the MediaElement instead of Shaka Payer correct?) 

On Chrome we end up still getting that segments that we do not need from the start of the stream. 

On Safari my test blows in cancellation of fetch requests.

In my opinion this feature does not need to depend on the vagaries of the MediaElement implementation.  I think it would be cleaner to use something other than the mediaElement.currentTime to convey updates to the start time. e.g. Player.setStartTime(startTime: number) and stash the value in the Player object.

**Comment by joeyparrish** [comment]
> I'm not having much luck with updating mediaElement.currentTime in the loadedmetadata event (which is emitted by the MediaElement instead of Shaka Payer correct?)

Yes, that's correct.

> In my opinion this feature does not need to depend on the vagaries of the MediaElement implementation. I think it would be cleaner to use something other than the mediaElement.currentTime to convey updates to the start time. e.g. Player.setStartTime(startTime: number) and stash the value in the Player object.

I'm open to a new method for that, but there's more to it.

The trouble is that for src= playback on Safari and iOS, that ultimately **must** be implemented by choosing the right time/event to set `mediaElement.startTime`.  If we can't figure out the right timing for setting the value on WebKit, we can't implement it correctly either in the application _or_ in the library.

**Comment by caridley** [comment]
I would like to see this converted into an enhancement request. I have already submitted a PR to implement the change https://github.com/google/shaka-player/pull/3491

**Comment by theodab** [comment]
Closed by PR #3491 (commit 16c1810b0321724b0daa44cf7f4736b0b0983153)
